test/extensions/tracers/datadog/agent_http_client_test.cc

```diff
@@ -460,6 +466,116 @@ TEST_F(DatadogAgentHttpClientTest, OnBeforeFinalizeUpstreamSpanIsANoOp) {
   client_.onBeforeFinalizeUpstreamSpan(null_span, nullptr);
 }
 
+TEST_F(DatadogAgentHttpClientTest, SkipReportIfCollectorClusterHasBeenRemoved) {
+  // Verify the effect of onClusterAddOrUpdate()/onClusterRemoval() on reporting logic,
+  // keeping in mind that they will be called both for relevant and irrelevant clusters.
+  using testing::DoAll;
+  using testing::Return;
+  using testing::WithArg;
+
+  NiceMock<Upstream::MockClusterManager>& cm = cluster_manager_.instance_;
+  Upstream::ClusterUpdateCallbacks* cluster_update_callbacks =
+      cluster_manager_.cluster_update_callbacks_;
+
+  {
+    // Simulate removal of the relevant cluster.
+    cluster_update_callbacks->onClusterRemoval("fake_cluster");
+
+    // Verify that no report will be sent.
+    EXPECT_CALL(cm.thread_local_cluster_, httpAsyncClient()).Times(0);
+    EXPECT_CALL(cm.thread_local_cluster_.async_client_, send_(_, _, _)).Times(0);
+
+    // Attempt to send a request.
+    const auto ignore = [](auto&&...) {};
+    datadog::tracing::Expected<void> result = client_.post(url_, ignore, "", ignore, ignore);
+    EXPECT_TRUE(result);
+
+    // Verify observability.
+    EXPECT_EQ(1U, stats_.reports_skipped_no_cluster_.value());
+    EXPECT_EQ(0U, stats_.reports_sent_.value());
+    EXPECT_EQ(0U, stats_.reports_dropped_.value());
+    EXPECT_EQ(0U, stats_.reports_failed_.value());
+  }
+
+  {
+    // Simulate addition of an irrelevant cluster.
+    NiceMock<Upstream::MockThreadLocalCluster> unrelated_cluster;
+    unrelated_cluster.cluster_.info_->name_ = "unrelated_cluster";
+    cluster_update_callbacks->onClusterAddOrUpdate(unrelated_cluster);
+
+    // Verify that no report will be sent.
+    EXPECT_CALL(cm.thread_local_cluster_, httpAsyncClient()).Times(0);
+    EXPECT_CALL(cm.thread_local_cluster_.async_client_, send_(_, _, _)).Times(0);
+
+    // Attempt to send a request.
+    const auto ignore = [](auto&&...) {};
+    datadog::tracing::Expected<void> result = client_.post(url_, ignore, "", ignore, ignore);
+    EXPECT_TRUE(result);
+
+    // Verify observability.
+    EXPECT_EQ(2U, stats_.reports_skipped_no_cluster_.value());
+    EXPECT_EQ(0U, stats_.reports_sent_.value());
+    EXPECT_EQ(0U, stats_.reports_dropped_.value());
+    EXPECT_EQ(0U, stats_.reports_failed_.value());
+  }
+
+  {
```

Comment:
Are these braced stanzas independent or in sequence?

If they are independent consider putting the 2 init lines in a helper function in the test fixture, and making each of these braced stanzas each be separate TEST_F methods.

Reply:
They're in sequence due to the accumulation of values in `stats_`.

This test is moved from https://github.com/envoyproxy/envoy/blob/7a1b9b5412420c2cd18c2d6a0c115eac0a94ed71/test/extensions/tracers/datadog/datadog_tracer_impl_test.cc#L265.

They could probably be separated into individual test methods with a little renovation. I'd prefer to leave it as-is, but if it smells, let me know.